homeassistant/components/tplink/binary_sensor.py

```diff
@@ -0,0 +1,111 @@
+"""Support for TPLink binary sensors."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Final
+
+from kasa import Device, Feature
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TPLinkConfigEntry
+from .coordinator import TPLinkDataUpdateCoordinator
+from .entity import (
+    CoordinatedTPLinkFeatureEntity,
+    TPLinkFeatureEntityDescription,
+    _description_for_feature,
+    _entities_for_device_and_its_children,
+)
+
+
+@dataclass(frozen=True, kw_only=True)
+class TPLinkBinarySensorEntityDescription(
+    BinarySensorEntityDescription, TPLinkFeatureEntityDescription
+):
+    """Base class for a TPLink feature based sensor entity description."""
+
+
+BINARYSENSOR_DESCRIPTIONS: Final = [
+    TPLinkBinarySensorEntityDescription(
+        key="overheated",
+        device_class=BinarySensorDeviceClass.HEAT,
+    ),
+    TPLinkBinarySensorEntityDescription(
+        key="battery_low",
+        device_class=BinarySensorDeviceClass.BATTERY,
+    ),
+    TPLinkBinarySensorEntityDescription(
+        key="cloud_connection",
+        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+    ),
+    # To be replaced & disabled per default by the upcoming update platform.
+    TPLinkBinarySensorEntityDescription(
+        key="update_available",
+        device_class=BinarySensorDeviceClass.UPDATE,
+    ),
+    TPLinkBinarySensorEntityDescription(
+        key="temperature_warning",
+        device_class=BinarySensorDeviceClass.PROBLEM,
+    ),
+    TPLinkBinarySensorEntityDescription(
+        key="humidity_warning", device_class=BinarySensorDeviceClass.PROBLEM
+    ),
+]
+
+BINARYSENSOR_DESCRIPTIONS_MAP = {desc.key: desc for desc in BINARYSENSOR_DESCRIPTIONS}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    children_coordinators = data.children_coordinators
+    device = parent_coordinator.device
+
+    entities = _entities_for_device_and_its_children(
```

Comment:
There is probably not need to make these protected inside the integration. We don't expect any other integrations import from tplink

Reply:
Yeah, at least some of the helpers should likely go into the base entity. Perhaps it's better to deal with the cleanup (as part of this PR?) before updating the other PRs.

Reply:
Either works. Its a simple change so it won't be hard to review

Reply:
Renamed to `entities_for_device_and_its_children` in https://github.com/home-assistant/core/pull/120064.
`_description_for_feature` and `category_for_feature` moved to classmethods in the same PR.